crates/ide-assists/src/handlers/merge_nested_if.rs

```diff
@@ -0,0 +1,235 @@
+use ide_db::syntax_helpers::node_ext::is_pattern_cond;
+use syntax::{
+    ast::{self, AstNode, BinaryOp},
+    T,
+};
+
+use crate::{
+    assist_context::{AssistContext, Assists},
+    AssistId, AssistKind,
+};
+// Assist: merge_nested_if
+//
+// This transforms if expressions of the form `if x { if y {A} }` into `if x && y {A}`
+// This assist can only be applied with the cursor on `if`.
+//
+// ```
+// fn main() {
+//    i$0f x == 3 { if y == 4 { 1 } }
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//    if x == 3 && y == 4 { 1 }
+// }
+// ```
+pub(crate) fn merge_nested_if(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let if_keyword = ctx.find_token_syntax_at_offset(T![if])?;
+    let expr = ast::IfExpr::cast(if_keyword.parent()?)?;
+    let if_range = if_keyword.text_range();
+    let cursor_in_range = if_range.contains_range(ctx.selection_trimmed());
+    if !cursor_in_range {
+        return None;
+    }
+
+    //should not apply to if with else branch.
+    if expr.else_branch().is_some() {
+        return None;
+    }
+
+    let cond = expr.condition()?;
+    //should not apply for if-let
+    if is_pattern_cond(cond.clone()) {
+        return None;
+    }
+
+    let cond_range = cond.syntax().text_range();
+
+    //check if the then branch is a nested if
+    let then_branch = expr.then_branch()?;
+    let stmt = then_branch.stmt_list()?;
+    if stmt.statements().count() != 0 {
+        return None;
+    }
+
+    let nested_if_to_merge = then_branch.syntax().descendants().find_map(ast::IfExpr::cast)?;
```

Comment:
```suggestion
    let nested_if_to_merge = then_branch.expr().and_then(|e| match e { ast::Expr::IfExpr(e) => Some(e), _ => None })?;
```

Reply:
The `then_branch` is of type `BlockExpr`, which does not have an `expr()` method. However, it has a `tail_expr()` method. Should I use it? It also passes all tests.

Reply:
Yes that is the one you want